rfc/0011-audit-checks.md

```diff
@@ -0,0 +1,354 @@
+
+## Creating a `PolicyReport` object
+
+The following section describes how, given a Namespace, we will produce a `PolicyReport` describing the compliance status.
+
+The process to create the `PolicyReport` for a single Namespace can be applied also to generate the `ClusterPolicyReport`, which is the one targeting only cluster-wide resources.
+
+### Step 1: create association between `ClusterAdmissionPolicy` and namespaces
+
+`ClusterAdmissionPolicy` by default inspect events happening inside of all the namespaces, however users can limit their scope
+by using the `spec.namespaceSelector` field.
+
+We need to create a dictionary that holds the following information:
+
+  * key: name of the `ClusterAdmissionPolicy`
+  * value: a list containing the names of the namespaces that are relevant to the `ClusterAdmissionPolicy`
+
+This operation can be done by iterating over each `ClusterAdmissionPolicy` and
+performing a "list namespaces" query against the API server that uses the
+same filter specified by the `ClusterAdmissionPolicy`.
+
+### Step 2: find relevant policies
+
+Given a Namespace to be audited, create a list of all the policies that are interested about the Namespace.
+
+All the `AdmissionPolicy` resources that are defined inside of the namespace are part of the list.
+
+For `ClusterAdmissionPolicy` resources, use the information stored inside of the
+dictionary that was created during the previous step.
+
+> Note: we will consider only policies that inspect `CREATE` events
+
+### Step 3: find the relevant Kubernetes resources
+
+Now we have a list of all the policies that are interested about the Namespace being audited.
+We iterate over each one of them and create a dictionary that has:
+
+* Key: the Kubernetes resource type, for example `Pod`
+* Value: a list of policies that evaluate the given Kubernetes resource
+
+The idea is to have a map of policies that tell us these information:
+
+* `Pod` resources are relevant to `polA`, `polB`
+* `Deployment` resources are relevant to `polA`
+* `Ingress` resources are relevant to `polC`
+
+### Step 4: fetch Kubernetes resources to be audited
+
+Next we iterate over the keys of this dictionary and, for each one of them we
+query the API server to get all the resources of type `X` defined inside of our Namespace.
+
+For example, given the previous example, we would end up with:
+
+* List of all the `Pod` objects defined inside of the audited Namespace
+* List of all the `Deployment` objects defined inside of the audited Namespace
+* List of all the `Ingress` objects defined inside of the audited Namespace
+
+**Note:** there's no need to do any caching of the response we get from the Kubernetes API server because of the following reasons:
+
+* We work on a per Namespace basis, these query results are useful only in the context of the Namespace being audited
+* The audit reports is generated every X minutes (maybe 10/30/60 minutes), cached data would be useless after this time
+
```

Comment:
We need to take into account the `objectSelector` before starting with Step5.
The `objectSelector` is of type `metav1.LabelSelector` same as the `namespaceSelector`. That means we can use the same method [FormatLabelSelector](https://github.com/kubernetes/apimachinery/blob/master/pkg/apis/meta/v1/helpers.go#L169) to convert it to the type we need for doing the filtering.

We need to check if the resource matches the `objectSelector` before moving to step 5, ignore the resource if it doesn't match the `objectSelector`. This can be done using the [Matches](https://github.com/kubernetes/apimachinery/blob/master/pkg/labels/selector.go#L51) method which doesn't make any call to the apiServer.

Reply:
The goal of this step is:

* Given a Namespace
* Given a list of Kubernetes resource types that are relevant to the policies to be evaluated,
* Get a list of all the Kubernetes instances defined inside of that Namespace

For example, you know you have 2 policies interested about `Pod` -> we get all the `Pod` instances defined inside of the Namespace. Once of the policies might have an `objectSelector`, but we don't care about that right now.

The filtering can be done during step 5, if this doesn't require extra calls to the API server. Which seems the case.
See below